internal/resourcestore/resourcestore.go

```diff
@@ -18,13 +18,36 @@ const sleepTimeBeforeCleanup = 1 * time.Minute
 // Another routine can request a watcher for a resource by calling WatcherForResource.
 // All watchers will be notified when the resource has successfully been created.
 type ResourceStore struct {
-	resources map[string]*Resource
-	timeout   time.Duration
-	closeChan chan struct{}
-	closed    bool
-	sync.Mutex
+	resources    map[string]*Resource
+	stages       map[string]Stage
+	timeout      time.Duration
+	closeChan    chan struct{}
+	closed       bool
+	resourceLock sync.Mutex
+	stageLock    sync.Mutex
```

Comment:
my current thouht is separating the locking mechanisms will reduce contention. still, if there are hundreds of containers or pods being created then this will get very busy. may need something sharded by key so each item can be updated atomically without bothering the others...

Reply:
migrating to a RWLock might have some benefit.

Reply:
unfortunately, this value is write often, and rarely read. I have thought about using a RW mutex on the whole map and only adding names when we have a write lock, and then having an [atomic value](https://pkg.go.dev/sync/atomic#Value) for the Stage...